bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/auth/client/oauth2/StorageCipher.java

```diff
@@ -0,0 +1,55 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.auth.client.oauth2;
+
+import java.security.GeneralSecurityException;
+
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * This allows the encryption and decryption to be performed before saving to storage.
+ *
+ * @author Gary Tse - initial contribution
+ *
+ */
+public interface StorageCipher {
+
+    /**
+     * A unique cipher identifier per each implementation of StorageCipher.
+     * It allows the OAuthStoreHandler to choose which cipher implementation to use.
+     * This is particularly important when old ciphers becomes out-dated and
+     * need to be replaced by new implementations.
+     *
+     * @return unique identifier
+     */
+    public String getUniqueCipherId();
+
+    /**
+     * Encrypt the plainText, then produce a base64 encoded cipher text
+     *
+     * @param plainText
+     * @return base64-encoded( encrypted( text ) )
+     * @throws GeneralSecurityException
+     */
+    public String encrypt(@Nullable String plainText) throws GeneralSecurityException;
+
+    /**
+     * Decrypt the base64 encoded cipher text.
+     *
+     * @param base64CipherText This should be the result from the {@link #encrypt(String)}
+     * @return plain text
+     * @throws GeneralSecurityException
```

Comment:
Shouldn't the exception be described when it would be thrown?

Reply:
GeneralSecurityException is from java.security.GeneralSecurityException.
This is like documenting "@throw IOException" on a method that is writing to a file.  

IMHO there is little added value to put extra comment here when Javadoc already explained all that.

Reply:
According to the general programming practice as followed by Java Language Designers, I have not seen any method specifying only "@throws IOException" without any description. If a method if writing to a file, and it throws IOException, it is clearly added to javadoc with the proper message.

From the javadoc of `FileInputStream#read(..)`
```java
/**
     * Reads up to <code>len</code> bytes of data from this input stream
     * into an array of bytes. If <code>len</code> is not zero, the method
     * blocks until some input is available; otherwise, no
     * bytes are read and <code>0</code> is returned.
     *
     * @param      b     the buffer into which the data is read.
     * @param      off   the start offset in the destination array <code>b</code>
     * @param      len   the maximum number of bytes read.
     * @return     the total number of bytes read into the buffer, or
     *             <code>-1</code> if there is no more data because the end of
     *             the file has been reached.
     * @exception  NullPointerException If <code>b</code> is <code>null</code>.
     * @exception  IndexOutOfBoundsException If <code>off</code> is negative,
     * <code>len</code> is negative, or <code>len</code> is greater than
     * <code>b.length - off</code>
     * @exception  IOException  if an I/O error occurs.
     */
    public int read(byte b[], int off, int len) throws IOException {
        return readBytes(b, off, len);
    }
```